File: Javascript/src/01.AlgorithmicToolbox/Week5DynamicProgramming1/longestCommonSubsequenceThree.js
/**
 * Problem 5 - Longest Common Subsequence of Three Sequence
 * Compute the length of a longest common subsequence of three sequences
 * @param {Array.<string>} first first sequence
 * @param {Array.<string>} second second sequence
 * @param {Array.<string>} third thid sequence
 * @returns {number} length of longest common subsequence
 */
export function longestCommonSubsequenceThree(first, second, third) {
  const longestSubsequenceLengths = new Array(first.length + 1)
    .fill(0)
    .map(() =>
      new Array(second.length + 1)
        .fill(0)
        .map(() => new Array(third.length + 1).fill(0))
    );

  for (let x = 1; x <= first.length; ++x) {
    for (let y = 1; y <= second.length; ++y) {
      for (let z = 1; z <= third.length; ++z) {
        if (first[x - 1] === second[y - 1] && second[y - 1] === third[z - 1]) {
          // add 1 to existing subsequence
          longestSubsequenceLengths[x][y][z] =
            longestSubsequenceLengths[x - 1][y - 1][z - 1] + 1;
        } else {
          // max of prefixes
          longestSubsequenceLengths[x][y][z] = Math.max(
            longestSubsequenceLengths[x - 1][y][z],
            longestSubsequenceLengths[x][y - 1][z],
            longestSubsequenceLengths[x][y][z - 1]
          );
        }
      }
    }
  }
  return longestSubsequenceLengths[first.length][second.length][third.length];
}

File: Rust/src/_01_algorithmic_toolbox/week5_dynamic_programming1/longest_common_subsequence_three.rs
/// Problem 5 - Longest Common Subsequence of Three Sequence
/// Compute the length of a longest common subsequence of three sequences
pub fn longest_common_subsequence_three(first: &[u64], second: &[u64], third: &[u64]) -> u64 {
    let mut longest_common_subsequence_lengths =
        vec![vec![vec![0; third.len() + 1]; second.len() + 1]; first.len() + 1];

    for x in 1..=first.len() {
        for y in 1..=second.len() {
            for z in 1..=third.len() {
                if first[x - 1] == second[y - 1] && second[y - 1] == third[z - 1] {
                    // add 1 to existing subsequence
                    longest_common_subsequence_lengths[x][y][z] =
                        longest_common_subsequence_lengths[x - 1][y - 1][z - 1] + 1;
                } else {
                    // max of prefixes
                    longest_common_subsequence_lengths[x][y][z] = std::cmp::max(
                        std::cmp::max(
                            longest_common_subsequence_lengths[x - 1][y][z],
                            longest_common_subsequence_lengths[x][y - 1][z],
                        ),
                        longest_common_subsequence_lengths[x][y][z - 1],
                    )
                }
            }
        }
    }
    longest_common_subsequence_lengths[first.len()][second.len()][third.len()]
}

#[cfg(test)]
mod test {
    use super::*;
    #[test]
    fn return_longest_common_subsequence_of_two_sequences() {
        assert_eq!(
            longest_common_subsequence_three(&vec![1, 2, 3], &vec![2, 1, 3], &vec![1, 3, 5]),
            2,
        );
        assert_eq!(
            longest_common_subsequence_three(
                &vec![8, 3, 2, 1, 7],
                &vec![8, 2, 1, 3, 8, 10, 7],
                &vec![6, 8, 3, 1, 4, 7]
            ),
            3,
        );
    }
}
